library/ssl_tls13_client.c

```diff
@@ -2120,6 +2120,34 @@ static int ssl_parse_server_psk_identity_ext( mbedtls_ssl_context *ssl,
 
 #endif
 
+/* Early Data Extension
+*
+* struct {} Empty;
+*
+* struct {
+*   select (Handshake.msg_type) {
+*     case new_session_ticket:   uint32 max_early_data_size;
+*     case client_hello:         Empty;
+*     case encrypted_extensions: Empty;
+*   };
+* } EarlyDataIndication;
+*/
+int ssl_parse_early_data_ext(mbedtls_ssl_context *ssl,
+    const unsigned char *buf, size_t len) {
+  ( ( void )buf );
+  if (ssl->state == MBEDTLS_SSL_ENCRYPTED_EXTENSIONS
+      && ssl->handshake->early_data == MBEDTLS_SSL_EARLY_DATA_ON
+      && len == 0) {
+    ssl->early_data_status = MBEDTLS_SSL_EARLY_DATA_ACCEPTED;
+    return 0;
+  }
+  return MBEDTLS_ERR_SSL_INTERNAL_ERROR;
+}
+
+int mbedtls_ssl_get_early_data_status(mbedtls_ssl_context *ssl)
+{
+  return ssl->early_data_status;
```

Comment:
The documentation for this function says that it must not be called prior to handshake completion. Shall we check this here?

There's always a trade-off between protecting the user from itself, and saving code-size, and for Mbed TLS I tend to give priority to code-size. However, when it comes to high-level state checks such as "Handshake done?", I think we tend to do those checks in the code, so for consistency I think it should be added here, too.

Also, should we fail if this is called by the server?

Reply:
We used macros, `MBEDTLS_SSL_CLI_C `, for this call, so it won't be used by server. We also removed the unnecessary set of `MBEDTLS_SSL_EARLY_DATA_NOT_SENT`. We should be able use the plain value after simplify those cases.